File: rock/sdk/common/exceptions.py
from rock._codes import codes
from rock.actions import CommitErrorCode, SandboxResponse
from rock.sdk.common.e2b import E2BConnectCode
from rock.utils.deprecated import deprecated


class WorkerCommitError(Exception):
    def __init__(self, code: CommitErrorCode, message: str):
        super().__init__(message)
        self.code = code


class E2BSandboxNotFoundError(Exception):
    pass


class E2BConnectError(Exception):
    def __init__(self, code: E2BConnectCode, message: str) -> None:
        super().__init__(message)
        self.code = code
        self.message = message


class RockException(Exception):
    _code: codes = None

    def __init__(self, message, code: codes = None):
        super().__init__(message)
        self._code = code

    @property
    def code(self):
        return self._code


@deprecated("This exception is deprecated")
class InvalidParameterRockException(RockException):
    def __init__(self, message):
        super().__init__(message)


class BadRequestRockError(RockException):
    def __init__(self, message, code: codes = codes.BAD_REQUEST):
        super().__init__(message, code)


class SandboxNotFoundRockError(BadRequestRockError):
    pass


class InternalServerRockError(RockException):
    def __init__(self, message, code: codes = codes.INTERNAL_SERVER_ERROR):
        super().__init__(message, code)


class CommandRockError(RockException):
    def __init__(self, message, code: codes = codes.COMMAND_ERROR):
        super().__init__(message, code)


def raise_for_code(code: codes, message: str):
    if code is None or codes.is_success(code):
        return

    if codes.is_client_error(code):
        raise BadRequestRockError(message)
    if codes.is_server_error(code):
        raise InternalServerRockError(message)
    if codes.is_command_error(code):
        raise CommandRockError(message)

    raise RockException(message, code=code)


def from_rock_exception(e: RockException) -> SandboxResponse:
    return SandboxResponse(code=e.code, failure_reason=str(e))

File: rock/ts-sdk/src/common/exceptions.ts
/**
 * ROCK Exception classes
 */

import { Codes } from '../types/codes.js';
import { SandboxResponse } from '../types/responses.js';

/**
 * Base ROCK exception
 */
export class RockException extends Error {
  protected _code: Codes | null = null;

  constructor(message: string, code?: Codes) {
    super(message);
    this.name = 'RockException';
    this._code = code ?? null;
  }

  get code(): Codes | null {
    return this._code;
  }
}

/**
 * Invalid parameter exception (deprecated)
 * @deprecated Use BadRequestRockError instead
 */
export class InvalidParameterRockException extends RockException {
  constructor(message: string) {
    super(message);
    this.name = 'InvalidParameterRockException';
  }
}

/**
 * Bad request error (4xxx)
 */
export class BadRequestRockError extends RockException {
  constructor(message: string, code: Codes = Codes.BAD_REQUEST) {
    super(message, code);
    this.name = 'BadRequestRockError';
  }
}

/**
 * Internal server error (5xxx)
 */
export class InternalServerRockError extends RockException {
  constructor(message: string, code: Codes = Codes.INTERNAL_SERVER_ERROR) {
    super(message, code);
    this.name = 'InternalServerRockError';
  }
}

/**
 * Command execution error (6xxx)
 */
export class CommandRockError extends RockException {
  constructor(message: string, code: Codes = Codes.COMMAND_ERROR) {
    super(message, code);
    this.name = 'CommandRockError';
  }
}

/**
 * Raise appropriate exception based on status code
 */
export function raiseForCode(code: Codes | null | undefined, message: string): void {
  if (code === null || code === undefined || isSuccessCode(code)) {
    return;
  }

  if (isClientErrorCode(code)) {
    throw new BadRequestRockError(message, code);
  }
  if (isServerErrorCode(code)) {
    throw new InternalServerRockError(message, code);
  }
  if (isCommandErrorCode(code)) {
    throw new CommandRockError(message, code);
  }

  throw new RockException(message, code);
}

/**
 * Convert RockException to SandboxResponse
 */
export function fromRockException(e: RockException): SandboxResponse {
  return {
    code: e.code ?? undefined,
    exitCode: 1,
    failureReason: e.message,
  };
}

// Helper functions for code checking
function isSuccessCode(code: Codes): boolean {
  return code >= 2000 && code <= 2999;
}

function isClientErrorCode(code: Codes): boolean {
  return code >= 4000 && code <= 4999;
}

function isServerErrorCode(code: Codes): boolean {
  return code >= 5000 && code <= 5999;
}

function isCommandErrorCode(code: Codes): boolean {
  return code >= 6000 && code <= 6999;
}
